python/src/nanoarrow/schema.py

```diff
@@ -0,0 +1,448 @@
+
+    UNINITIALIZED = CArrowType.UNINITIALIZED
+    NA = CArrowType.NA
+    BOOL = CArrowType.BOOL
+    UINT8 = CArrowType.UINT8
+    INT8 = CArrowType.INT8
+    UINT16 = CArrowType.UINT16
+    INT16 = CArrowType.INT16
+    UINT32 = CArrowType.UINT32
+    INT32 = CArrowType.INT32
+    UINT64 = CArrowType.UINT64
+    INT64 = CArrowType.INT64
+    HALF_FLOAT = CArrowType.HALF_FLOAT
+    FLOAT = CArrowType.FLOAT
+    DOUBLE = CArrowType.DOUBLE
+    STRING = CArrowType.STRING
+    BINARY = CArrowType.BINARY
+    FIXED_SIZE_BINARY = CArrowType.FIXED_SIZE_BINARY
+    DATE32 = CArrowType.DATE32
+    DATE64 = CArrowType.DATE64
+    TIMESTAMP = CArrowType.TIMESTAMP
+    TIME32 = CArrowType.TIME32
+    TIME64 = CArrowType.TIME64
+    INTERVAL_MONTHS = CArrowType.INTERVAL_MONTHS
+    INTERVAL_DAY_TIME = CArrowType.INTERVAL_DAY_TIME
+    DECIMAL128 = CArrowType.DECIMAL128
+    DECIMAL256 = CArrowType.DECIMAL256
+    LIST = CArrowType.LIST
+    STRUCT = CArrowType.STRUCT
+    SPARSE_UNION = CArrowType.SPARSE_UNION
+    DENSE_UNION = CArrowType.DENSE_UNION
+    DICTIONARY = CArrowType.DICTIONARY
+    MAP = CArrowType.MAP
+    EXTENSION = CArrowType.EXTENSION
+    FIXED_SIZE_LIST = CArrowType.FIXED_SIZE_LIST
+    DURATION = CArrowType.DURATION
+    LARGE_STRING = CArrowType.LARGE_STRING
+    LARGE_BINARY = CArrowType.LARGE_BINARY
+    LARGE_LIST = CArrowType.LARGE_LIST
+    INTERVAL_MONTH_DAY_NANO = CArrowType.INTERVAL_MONTH_DAY_NANO
+
+    def __arrow_c_schema__(self):
+        # This will only work for parameter-free types
+        c_schema = CSchemaBuilder.allocate().set_type(self.value).finish()
+        return c_schema._capsule
+
+
+class TimeUnit(enum.Enum):
+    """Unit enumerator for timestamp, duration, and time types."""
+
+    SECOND = CArrowTimeUnit.SECOND
+    MILLI = CArrowTimeUnit.MILLI
+    MICRO = CArrowTimeUnit.MICRO
+    NANO = CArrowTimeUnit.NANO
+
+
+class Schema:
+    """The Schema is nanoarrow's data type representation, encompasing the role
+    of PyArrow's ``Schema``, ``Field``, and ``DataType``. This scope maps to
+    that of the ArrowSchema in the Arrow C Data interface."""
+
+    def __init__(
+        self,
+        obj,
+        *,
+        nullable=None,
+        **params,
+    ) -> None:
+        """Create Schema objects
+
+        Create Schema objects from (1) a :class:`Type` or (1) any object supported by
+        :func:`c_schema`. When creating a class from a  a :class:`Type`, some
+        parameters may be required.
+
+        >>> import nanoarrow as na
+        >>> import pyarrow as pa
+        >>> schema = na.Schema(na.Type.INT32)
+        >>> schema = na.Schema(na.Type.DURATION, unit=na.TimeUnit.SECOND)
+        >>> schema = na.Schema(pa.int32())
+        """
+
+        if isinstance(obj, Type):
+            self._c_schema = _c_schema_from_type_and_params(obj, params, nullable)
+        elif not params:
+            self._c_schema = c_schema(obj)
```

Comment:
The `nullable` keyword is being ignored here. Should we check `and nullable is None` in addition, so we raise an error when set and passing an object that already has that information encoded:

```
In [36]: na.Schema(pa.int16()).nullable
Out[36]: True

In [37]: na.Schema(pa.int16(), nullable=False).nullable
Out[37]: True

In [38]: na.Schema(pa.field("test", pa.int16(), nullable=False), nullable=True).nullable
Out[38]: False
```